### PR TITLE
Load narrowKlass from header, AArch64 assembler implementation

### PR DESCRIPTION
**Pull request by rkennke**
This implements MacroAssembler::load_klass() to load the (narrow)Klass* from object header. Just like the x86_64 implementation, it checks whether it can take the fast path (object unlocked -> load from upper 32bits of header), or else calls the runtime to get a stable header and load from that.

It adds a runtime call stub, which will also be used in the C2 implementation. It also adds nklass_offset_in_bytes() which will also be used in C2 impl. The part in generate_verify_oop() is a little nasty, I added a comment that explains what's going on.

Testing:
 - [x] tier1 (aarch64)
 - [x] tier2 (aarch64)
 - [x] tier3 (aarch64)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/lilliput pull/36/head:pull/36` \
`$ git checkout pull/36`

Update a local copy of the PR: \
`$ git checkout pull/36` \
`$ git pull https://git.openjdk.java.net/lilliput pull/36/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 36`

View PR using the GUI difftool: \
`$ git pr show -t 36`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/lilliput/pull/36.diff">https://git.openjdk.java.net/lilliput/pull/36.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/oops/oop.hpp
Comment (shipilev) [inline src/hotspot/share/oops/oop.hpp:311]: Ew. Need to assert that `markWord::klass_shift` is actually a multiple of 8.

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
Comment (shipilev) [inline src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:596]: Something should assert that `oopDesc::nklass_offset_in_bytes()` does not break the alignment. Probably assert that `oopDesc::nklass_offset_in_bytes()` is a multiple of 4?

**Review by shipilev — COMMENTED** [review]
I'll defer to Andrew to examine AArch64 code, but I have two quick comments.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3718]:
```suggestion
  tbnz(tmp, exact_log2(markWord::lock_mask_in_place, slow);
```

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3722]:
```suggestion
  lsr(dst, tmp, markWord::klass_shift);
```

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3737]:
```suggestion
  far_call(RuntimeAddress(StubRoutines::load_nklass()), rscratch1);
```

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:597]: `ldrw()` will definitely complain if it isn't.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3733]: This is all very weird, and looks confusing. I think it might be safe to force the callers of `load_klass()` to provide a temp.

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:6621]:
```suggestion
    __ pop_call_clobbered_registers_except(r0);
```

File: src/hotspot/share/oops/oop.hpp
Comment (theRealAph) [inline src/hotspot/share/oops/oop.hpp:312]:
```suggestion
    STATIC_ASSERT(markWord::klass_shift % 8 == 0, "klass_shift must be multiple of 8");
```

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:6619]:
```suggestion
    __ push_call_clobbered_registers_except(r0);
```
... everywhere.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3718]: Oh, sorry. Right you are.

File: src/hotspot/share/oops/oop.hpp
Comment (rkennke) [reply src/hotspot/share/oops/oop.hpp]: Done that.

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
Comment (rkennke) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:597]: Done that.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (rkennke) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3722]: Done.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (rkennke) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3737]: Done (without the rscratch1 arg, which is the default anyway).

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (rkennke) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3733]: I checked again, and it actually looks like we don't need to preserve rscratch1 and rscratch2. I added asserts to verify that src register isn't one of the rscratch registers and would risk to get clobbered. It doesn't matter if dst is rscratch register.

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
Comment (rkennke) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:6621]: Done.

File: src/hotspot/share/oops/oop.hpp
Comment (rkennke) [reply src/hotspot/share/oops/oop.hpp:312]: Done.

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
Comment (rkennke) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:6619]: Done.

**Comment by rkennke** [comment]
@shipilev @theRealAph Thanks for reviewing! Please approve if you are ok with this updated PR.
Thanks, Roman

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3733]: To what does this comment refer?

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3720]:
```suggestion
  andr(dst, tmp, markWord::lock_mask_in_place);
  cbnz(dst, Assembler::NE, slow);
```

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:6620]: I'm tempted to suggest saving flags here too. But I guess you've checked all the places that use `load_klass()`. At least the comment for `load_klass()` should advertise exactly what gets clobbered.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (rkennke) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3720]: I believe that wouldn't work either. The fast path that is following this code requires that the upper 32 bits (the narrowKlass) of tmp remain intact: it shifts them to the right by 32bits and decodes the narrowKlass.

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
Comment (rkennke) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:6620]: Hmm yeah, maybe. I found one instance where flags are live around this call, but could work around it:

https://github.com/rkennke/lilliput/blob/86c13e4dfc6c1e53c9c97fc686d299c4615fd28d/src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp#L587

Also, it's not only the stub that clobbers the condition flags, but the fast-path, too.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3720]: > I believe that wouldn't work either. The fast path that is following this code requires that the upper 32 bits (the narrowKlass) of tmp remain intact: it shifts them to the right by 32bits and decodes the narrowKlass.

So what? tmp is untouched.

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:6620]: Fast path is fixable.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp
Comment (rkennke) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3719]: > > I believe that wouldn't work either. The fast path that is following this code requires that the upper 32 bits (the narrowKlass) of tmp remain intact: it shifts them to the right by 32bits and decodes the narrowKlass.
> 
> So what? tmp is untouched.

We don't guarantee that tmp != dst. We only guarantee that tmp != src to avoid clobbering src.

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3719]: OK, I give up. Please document 'clobbers flags' very clearly.

Comment (rkennke) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3719]: > OK, I give up. Please document 'clobbers flags' very clearly.

I will try to make it that the caller ensures that src != dst, and also passes a tmp register, then we can avoid clobbering the flags register altogether and also avoid the register shuffling madness. Not sure if that's always feasable, though.

Comment (rkennke) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp]: Alright, ensuring src != dst is relatively easy and seems worth doing (makes fast-path faster). I tried using rscratch2 as tmp register and use the proposed andr/cbnz sequence, but apparently rscratch2 is live in at least one path, and I could not easily find it. I documented the register usage, and that the load_klass() clobbers condition flags. That seems good enough for now, especially considering that at some point in the future, it will all go away (when locking is not overlaying the Klass* stuff in the header anymore).
@theRealAph ok now?

**Comment by rkennke** [comment]
I pushed more changes:
 - Split load_klass() and load_nklass() (we need the nklass version separately in a few places)
 - Cleanup remaining direct loads from klass_offset_in_bytes() in aarch64, replacing them with load_[n]klass()
 - Pass correct offset to null-checks that are preceding load_klass()
 - Trim code size estimate

@theRealAph and/or @shipilev please re-review? Thanks!

**Review by theRealAph — APPROVED** [review]
It looks good now.
Having said that, the convention is mostly that macros can use rscratchN registers freely internally, and that any caller needs to be aware of that. 
Here is the opposite of that convention: `load_nklass` carefully avoids clobbering `rscratch1` and `rscratch2` in order to make things easier for the caller. So, while this is something of a surprise for the maintainer, it's an elegant way to solve the problem.

**Comment by rkennke** [comment]
> It looks good now. Having said that, the convention is mostly that macros can use rscratchN registers freely internally, and that any caller needs to be aware of that. Here is the opposite of that convention: `load_nklass` carefully avoids clobbering `rscratch1` and `rscratch2` in order to make things easier for the caller. So, while this is something of a surprise for the maintainer, it's an elegant way to solve the problem.

Thanks!
I believe the trouble is that load_klass() is used in some C1 generated code paths, and C1 register allocator doesn't ignore rscratch1 and rscratch2. I may be wrong, though.

**Comment by theRealAph** [comment]
> Thanks! I believe the trouble is that load_klass() is used in some C1 generated code paths, and C1 register allocator doesn't ignore rscratch1 and rscratch2. I may be wrong, though.

I hope you are! - None of the register allocators can touch the scratch registers. Lots of hand-written code depends on that.

**Comment by rkennke** [comment]
> > Thanks! I believe the trouble is that load_klass() is used in some C1 generated code paths, and C1 register allocator doesn't ignore rscratch1 and rscratch2. I may be wrong, though.
> 
> I hope you are! - None of the register allocators can touch the scratch registers. Lots of hand-written code depends on that.


> > Thanks! I believe the trouble is that load_klass() is used in some C1 generated code paths, and C1 register allocator doesn't ignore rscratch1 and rscratch2. I may be wrong, though.
> 
> I hope you are! - None of the register allocators can touch the scratch registers. Lots of hand-written code depends on that.

I had a look at the C1 register allocator, and could verify that I was wrong.

Thanks for reviewing and helping!

/integrate

**Pull request closed without merging**
